alibi/explainers/similarity/base.py

```diff
@@ -0,0 +1,171 @@
+from abc import ABC
+from typing import TYPE_CHECKING, Callable, Union, Tuple, Any
+
+import numpy as np
+from tqdm import tqdm
+
+from alibi.api.interfaces import Explainer
+from alibi.explainers.similarity.backends import _select_backend
+
+
+if TYPE_CHECKING:
+    import tensorflow
+    import torch
+
+
+class BaseSimilarityExplainer(Explainer, ABC):
+    """Base class for similarity explainers."""
+
+    def __init__(self,
+                 predictor: 'Union[tensorflow.keras.Model, torch.nn.Module]',
+                 loss_fn: '''Union[Callable[[tensorflow.Tensor, tensorflow.Tensor], tensorflow.Tensor],
+                                   Callable[[torch.Tensor, torch.Tensor], torch.Tensor]]''',
+                 sim_fn: Callable[[np.ndarray, np.ndarray], np.ndarray],
+                 store_grads: bool = False,
+                 seed: int = 0,
+                 backend: str = "tensorflow",
+                 device: str = "cpu",
+                 **kwargs
+                 ):
+        """Constructor
+
+        Parameters
+        ----------
+        predictor:
+            Model to be explained.
+        loss_fn:
+            Loss function.
+        sim_fn:
+            Similarity function. Takes two inputs and returns a similarity value.
+        store_grads:
+            Whether to precompute and store the gradients when fitting.
+        seed:
+            Seed to be used,
+        backend:
+            Deep learning backend: ``'tensorflow'`` | ``'torch'``. Default ``'tensorflow'``.
+        device:
+            Device to be used. Default `cpu`.
+        """
+
+        if backend not in ['torch', 'tensorflow']:
+            raise ValueError(f'Unknown backend {backend}. Consider using: `torch` | `tensorflow` .')
+
+        # Select backend.
+        self.backend = _select_backend(backend, **kwargs)
+        self.backend.set_device(device)
+
+        # Set seed for reproducibility.
+        self.backend.set_seed(seed)
+
+        self.predictor = predictor
+        self.loss_fn = loss_fn
+        self.sim_fn = sim_fn
+        self.store_grads = store_grads
+        self.seed = seed
+
+        super().__init__(**kwargs)
```

Comment:
I think the `Explainer` base class doesn't use any of these kwargs? The only thing it should take is the `meta` dictionary, so I'm a bit unsure how this works currently.

Reply:
Ah, I missed that `meta` is passed from above, but doesn't appear in the constructor signature. Perhaps being more explicit would help.